_posts/2018-08-14-grpc-with-protobuf.md

```diff
@@ -0,0 +1,225 @@
+    byte[] key;
+    byte[] value;
+  }
+
+  static final class CreateResponse {
+  }
+
+  static final class RetrieveRequest {
+    byte[] key;
+  }
+
+  static final class RetrieveResponse {
+    byte[] value;
+  }
+
+  static final class UpdateRequest {
+    byte[] key;
+    byte[] value;
+  }
+
+  static final class UpdateResponse {
+  }
+
+  static final class DeleteRequest {
+    byte[] key;
+  }
+
+  static final class DeleteResponse {
+  }
+```
+
+Because GSON uses reflection to determine how the fields in our classes map to the serialized JSON, we don't need to annotate the messages.
+
+Our client and server logic will use the request and response types, but gRPC needs to know how to produce and consume these messages.  To do this, we need to implement a [`Marshaller`](https://grpc.io/grpc-java/javadoc/io/grpc/MethodDescriptor.Marshaller.html).  A marshaller knows how to convert from an arbitrary type to an `InputStream`, which is then passed down into the gRPC core library.  It is also capable of doing the reverse transformation when decoding data from the network.  For GSON, here is what the marshaller looks like:
+
+```java
+  static <T> Marshaller<T> marshallerFor(Class<T> clz) {
+    Gson gson = new Gson();
+    return new Marshaller<T>() {
+      @Override
+      public InputStream stream(T value) {
+        return new ByteArrayInputStream(gson.toJson(value, clz).getBytes(StandardCharsets.UTF_8));
+      }
+
+      @Override
+      public T parse(InputStream stream) {
+        return gson.fromJson(new InputStreamReader(stream, StandardCharsets.UTF_8), clz);
+      }
+    };
+  }
+```
+
+Given a `Class` object for a some request or response, this function will produce a marshaller.  Using the marshallers, we can compose a full `MethodDescriptor` for each of the four CRUD methods.  Here is an example of the Method descriptor for _Create_:
+
+```java
+  static final MethodDescriptor<CreateRequest, CreateResponse> CREATE_METHOD =
+      MethodDescriptor.newBuilder(
+          marshallerFor(CreateRequest.class),
+          marshallerFor(CreateResponse.class))
+          .setFullMethodName(
+              MethodDescriptor.generateFullMethodName(SERVICE_NAME, "Create"))
+          .setType(MethodType.UNARY)
+          .build();
+```
+
+Note that if we were using Protobuf, we would use the existing Protobuf marshaller, and the method descriptors would be generated automatically.
+
+## Sending RPCs
+
+We need to update our [`KvClient`](https://github.com/carl-mastrangelo/kvstore/blob/b225d28c7c2f3c356b0f3753384b3329f2ab5911/src/main/java/io/grpc/examples/KvClient.java#L98), the gRPC client used in the previous post, to use our MethodDescriptors.  Additionally, since we won't be using any Protobuf types, the code needs to use `ByteBuffer` rather than `ByteString`.  That said, we can still use the `grpc-stub` package on Maven to issue the RPC.  Using the _Create_ method again as an example, here's how to make an RPC:
```

Comment:
Now that we can marshall JSON requests and responses, we can update our...

Reply:
Done.